### PR TITLE
ARROW-1142: [C++] Port over compression toolchain and interfaces from parquet-cpp, use Arrow-style error handling

### PR DESCRIPTION
**Pull request by wesm**

### CODE REVIEW

**Comment by wesm** [comment]
Appveyour build: https://ci.appveyor.com/project/wesm/arrow/build/1.0.535

**Comment by wesm** [comment]
Appveyor build: https://ci.appveyor.com/project/wesm/arrow/build/1.0.540. Now just waiting on Travis CI

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -0,0 +1,640 @@
...
+  set(JEMALLOC_HOME "$ENV{ARROW_BUILD_TOOLCHAIN}")
+  set(GFLAGS_HOME "$ENV{ARROW_BUILD_TOOLCHAIN}")
+  set(SNAPPY_HOME "$ENV{ARROW_BUILD_TOOLCHAIN}")
+  set(ZLIB_HOME "$ENV{ARROW_BUILD_TOOLCHAIN}")
+  set(BROTLI_HOME "$ENV{ARROW_BUILD_TOOLCHAIN}")
+
+  if (NOT DEFINED ENV{BOOST_ROOT})
+    # Since we have to set this in the environment, we check whether
+    # $BOOST_ROOT is defined inside here
+    set(ENV{BOOST_ROOT} "$ENV{ARROW_BUILD_TOOLCHAIN}")
```
Comment (cpcloud) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: Is it possible to leave `BOOST_ROOT` out of this? My build environment uses the OS boost and conda stuff for everything else. I'm also happy to explicitly set `BOOST_ROOT` if we want to tie boost to the build toolchain.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+  find_library(PTHREAD_LIBRARY pthread)
+  message(STATUS "Found pthread: ${PTHREAD_LIBRARY}")
+endif()
+
+# ----------------------------------------------------------------------
+# Add Boost dependencies (code adapted from Apache Kudu (incubating))
+
+set(Boost_DEBUG TRUE)
+set(Boost_USE_MULTITHREADED ON)
+set(Boost_ADDITIONAL_VERSIONS
```
Comment (cpcloud) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: Possible to add `1.64.0`, and `1.64` here as well?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+    set(BOOST_FILESYSTEM_LIBRARY boost_filesystem_shared)
+  endif()
+else()
+  # Find static boost headers and libs
+  # TODO Differentiate here between release and debug builds
+  set(Boost_USE_STATIC_LIBS ON)
+  if (ARROW_BOOST_HEADER_ONLY)
+    find_package(Boost)
+  else()
+    find_package(Boost COMPONENTS system filesystem regex REQUIRED)
```
Comment (cpcloud) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: Should `regex` be in there?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 2 above)
Comment (wesm) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: This is moved code from the original root CMakeLists.txt, but sure

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+    set(GFLAGS_CMAKE_ARGS -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+                          -DCMAKE_INSTALL_PREFIX=${GFLAGS_PREFIX}
+                          -DBUILD_SHARED_LIBS=OFF
+                          -DBUILD_STATIC_LIBS=ON
+                          -DBUILD_PACKAGING=OFF
+                          -DBUILD_TESTING=OFF
+                          -BUILD_CONFIG_TESTS=OFF
+                          -DINSTALL_HEADERS=ON
+                          -DCMAKE_CXX_FLAGS=${GFLAGS_CMAKE_CXX_FLAGS})
+    if (CMAKE_VERSION VERSION_GREATER "3.2")
```
Comment (cpcloud) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: Can this `if` be removed?

Comment (cpcloud) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: (and any others that might also be around)

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (wesm) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: If you don't have boost installed in your toolchain, this is actually fine (I think). Does CMake fail to find Boost when you set `$ARROW_BUILD_TOOLCHAIN`?

**Review by cpcloud — APPROVED** [review]

**Comment by wesm** [comment]
Opened https://issues.apache.org/jira/browse/ARROW-1148 about raising the minimum CMake version to 3.2, can do that in a follow up patch

**Comment by wesm** [comment]
Here's the Appveyor build: https://ci.appveyor.com/project/wesm/arrow/build/1.0.546. Bombs away

File: .readthedocs.yml [hunk 5]
```diff
@@ -1,2 +1,19 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
```
Comment (xhochy) [inline .readthedocs.yml]: We should remove this file as we don't use readthedocs anymore.

**Pull request closed without merging**